hera_pspec/pspecdata.py

```diff
@@ -894,30 +937,120 @@ def get_H(self, key1, key2, sampling=False):
 
-            V_ab = tr(C E_a C E_b)
+            E^{12,a} = (1/2) R_1 Q^a R_2.
+
+        In principle, this could be used to actually estimate q-hat. In other
+        words, we could call this function to get E, and then sandwich it with
+        two data vectors to get q_a. However, this should be slower than the
+        implementation in q_hat. So this should only be used as a helper
+        method for methods such as get_unnormed_V.
+
+        Note for the future: There may be advantages to doing the
+        matrix multiplications separately
 
-        FIXME: Must check factor of 2 with Wick's theorem for complex vectors,
-        and also check expression for when x_1 != x_2.
 
+        Parameters
+        ----------
+        key1, key2 : tuples or lists of tuples
+            Tuples containing indices of dataset and baselines for the two 
+            input datavectors. If a list of tuples is provided, the baselines 
+            in the list will be combined with inverse noise weights.
+
+        Returns
+        -------
+        E : array_like, complex
+            Set of E matrices, with dimensions (Ndlys, Nfreqs, Nfreqs).
+
+        """
+
+        if self.spw_Ndlys == None:
+            raise ValueError("Number of delay bins should have been set"
+                             "by now! Cannot be equal to None")
+
+        E_matrices = np.zeros((self.spw_Ndlys, self.spw_Nfreqs, self.spw_Nfreqs),
+                               dtype=np.complex)
+        R1 = self.R(key1)
+        R2 = self.R(key2)
+        for dly_idx in range(self.spw_Ndlys):
+            E_matrices[dly_idx] = np.einsum('ij,jk,kl', R1, self.get_Q_alt(dly_idx), R2)
+
+        return 0.5 * E_matrices
+    
+    def get_unnormed_V(self, key1, key2, model='empirical'):
+        """
+        Calculates the covariance matrix for unnormed bandpowers (i.e., the q
+        vectors). If the data were real and x_1 = x_2, the expression would be
+        
+            V_ab = 2 tr(C E_a C E_b), where E_a = (1/2) R Q^a R
+
+        When the data are complex, the expression becomes considerably more
+        complicated. Define
+
+            E^{12,a} = (1/2) R_1 Q^a R_2
+            C^1 = Cov(x1)
+            C^2 = Cov(x2)
+            P^{12} = <x1 x2> - <x1><x2>
+            S^{12} = <x1^* x2^*> - <x1^*> <x2^*>
+
+        Then
+
+            V_ab = tr(E^{12,a} C^2 E^{21,b} C^1)
+                    + tr(E^{12,a} P^{21} E^{12,b *} S^{21})
+
+        Note that
+            E^{12,a}_{ij}.conj = E^{21,a}_{ji}
+
+        This function estimates C^1, C^2, P^{12}, and S^{12} empirically by default
+
+        Note for future: Although the V matrix should be Hermitian by construction,
+        in practice there are precision issues and the Hermiticity is violated at
+        ~ 1 part in 10^15. (Which is ~the expected roundoff error). If something
+        messes up, it may be worth investigating this more.
+
+        Note for the future: If this ends up too slow, Cholesky tricks can be
```

Comment:
It's conceptually not too complicated, but it can come at the cost of stability. If all the matrices are well-behaved, it's fine, but things can get a little scary if there are eigenvalues that are close to zero.